addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -267,61 +318,75 @@ var SnippetEditor = Widget.extend({
+                    $snippet: this.$snippetBlock,
+                    callback: function (editor, $snippet) {
+                        for (var i in editor.snippetOptionInstances) {
+                            editor.snippetOptionInstances[i].onRemove();
+                        }
+                        resolve();
+                    },
+                });
             });
-        });
 
-        this.trigger_up('go_to_parent', {$snippet: this.$target});
-        var $parent = this.$target.parent();
-        this.$target.find('*').addBack().tooltip('dispose');
-        this.$target.remove();
-        this.$el.remove();
-
-        var node = $parent[0];
-        if (node && node.firstChild) {
-            if (!node.firstChild.tagName && node.firstChild.textContent === ' ') {
-                node.removeChild(node.firstChild);
+            this.trigger_up('go_to_parent', {$snippet: this.$snippetBlock});
+            var $parent = this.$snippetBlock.parent();
+            this.$snippetBlock.find('*').addBack().tooltip('dispose');
+            // The snippet has to be removed directly in jquery as well for the
+            // synchronous bubbling up system to remove the empty parent.
+            // Ideally, we should simply call remove inside a mutations observer
+            // callback such that the removal is synchronous while the vdom is
+            // updated through observed mutations without calling remove twice.
+            this.$snippetBlock.remove();
+            await this.editorHelpers.remove(context, this.$snippetBlock[0]);
+            this.$el.remove();
+
+            var node = $parent[0];
+            if (node && node.firstChild) {
+                if (!node.firstChild.tagName && node.firstChild.textContent === ' ') {
+                    await this.editorHelpers.remove(context, node.firstChild);
+                }
             }
-        }
 
-        if ($parent.closest(':data("snippet-editor")').length) {
-            var editor = $parent.data('snippet-editor');
-            while (!editor) {
-                var $nextParent = $parent.parent();
-                if (isEmptyAndRemovable($parent)) {
-                    $parent.remove();
+            if ($parent.closest(':data("snippet-editor")').length) {
+                var editor = $parent.data('snippet-editor');
+                while (!editor) {
+                    var $nextParent = $parent.parent();
+                    if (isEmptyAndRemovable($parent)) {
+                        await this.editorHelpers.remove(context, this.$parent[0]);
+                    }
+                    $parent = $nextParent;
+                    editor = $parent.data('snippet-editor');
+                }
+                if (isEmptyAndRemovable($parent, editor)) {
+                    // TODO maybe this should be part of the actual Promise being
+                    // returned by the function ?
+                    await new Promise((resolve)=> {
```

Comment:
Makes sense indeed. We have a very strict linter configured (with auto-fix on save) for our project and it was giving us hundreds of errors on the Odoo side of the code. We had to disable auto-fix on save and I guess with the rush we forgot to take the time to setup our linter for Odoo and do a pass with it. Not that that excuses it but there you get the explanation. We should take a moment for linting before 14.1.

Reply:
@qsm-odoo I was not aware Odoo had (finally) enforced a common linter configuration for all js files. If you have it, could you please share a link to the configuration so that we can update our workspaces ?

Reply:
@dmo-odoo I assumed he meant this: https://github.com/odoo/odoo/wiki/Javascript-coding-guidelines
I was using it before but since we configured our own linting for jw, that one replaced it. Then I had to disable it when working in Odoo because our linting was stricter and showed hundreds of problems in Odoo's js according to jw's linting standards. In my memory, we talked about that at the time and decided we'd fix that and do a linter pass later.

Reply:
Well this wiki is wrong. I remember working on gantt for V13 without a linter so clearly this is newer than that. We might want to do a linting pass indeed.

Reply:
This was added in the wiki in August 2017 for version 11.0 so I don't know what is wrong about it, but as indicated, it is just a sample one :wink: The point here was to use _any_ linter as any configuration would at least warn about missing `;` or unused variables.

Reply:
Note: the sample linter configuration provided in the wiki is also made to respect the generic guidelines on the same wiki page (like the if-else structure, the strict comparisons, etc) and those guidelines were decided by the framework team for 11.0 as well :wink: